### DISCUSSION TITLE
Relationship problem  "When initializing mapper"

### DISCUSSION THREAD

**Discussion by ericPavone** [fastapi/sqlmodel#900] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class BaseEntity(SQLModel):
    """Class with base data from the entities"""

    id: int | None = Field(default=None, primary_key=True)
    data_creation: str
    data_udpate: str | None
    
    class Config:
        arbitrary_types_allowed = True



class Product(
    BaseEntity,
    table=True,
):
    """Class the products"""

    __tablename__ = "products"

    name: str = Field(default="", le=200)
    description: TEXT = Field(default="", sa_type=TEXT)
    product_code: str = Field(default="")

    production_country: str = Field(default="", index=True)
    production_city: str = Field(default="", index=True)

    csr_code: int = Field(default_factory=lambda: randint(100000, 999999))
    csr_image: LONGBLOB = Field(default=None,nullable=False, sa_type=LONGBLOB)

    user_id: int = Field(default=None, foreign_key="users.id")
    user: "User" = Relationship(back_populates="products")




class User(
    BaseEntity,
    table=True
):
    """Class the users"""

    __tablename__ = "users"

    name: str = Field(default="", le=200)
    email: str = Field(default="", le=200, index=True, unique=True)
    password: str = Field(le=500)

    products: list["Product"] = Relationship(back_populates="user")
```

### Description

at the frist query i get this message.
When initializing mapper Mapper[User(users)], expression "relationship("list['Product']")" seems to be using a generic class as the argument to relationship(); please state the generic argument using an annotation, e.g. "products: Mapped[list['Product']] = relationship()"

but only if i add the navigation properties, if i use only the id to create the relation, all works

### Operating System

macOS

### SQLModel Version

0.0.16

### Python Version

Python 3.12.2

**Comment by ericPavone** [comment]
here is the problem.

from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ... import ... 
    
    
to keep the type, idk why, all was broken. removing it, all works.

**Reply by iamyb** [reply]
Hello, what's your final solution here? Seems i met similar issue...

**Reply by ericPavone** [reply]
The solution is to not use 

"from future import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
from ... import ...
" 
with the navigation properties defined as Object EG:

    products: list[Product] <--- real object type = Relationship(back_populates="user")
    
    use instead: 
    
    products: list["Product"]   <--- string version  = Relationship(back_populates="user")

**Reply by kent-foodini** [reply, marked as answer]
To help those who find this and puzzle over the discussion:

To sort out massive errors with SQLModel / SQLAlchemy, the "from future import annotations" statement in the file will completely bork SQLAlchemy with errors leading you to think the relationships are based on generic types.

DO NOT USE from future import annotations AND SQL MODEL

I took this line out after hours of trying things, and SQLModel started working again.

**Reply by bilalbit** [reply]
Hello, I’m facing a similar issue, but the error I’m encountering is slightly different. Could you kindly take a look at the issue I’ve described here? Thank you for your help!  https://github.com/fastapi/sqlmodel/discussions/1292

**Comment by RickBarretto** [comment]
I've been spending days trying to solve this problem. Hopefully I found this discussion.
Thanks @ericPavone and @kent-foodini!

**Comment by Pi3r2Dev** [comment]
Thanks a lot to @ericPavone and @kent-foodini !!

**Comment by nikolaysm** [comment]
What fixed it for me (SQLAlchemy 2.0 typing) was adding `sa_relationship` while also using `Mapped[].` 
For your example, that becomes:
```py
from __future__ import annotations
from sqlalchemy.orm import Mapped, relationship
from sqlmodel import Field, Relationship, SQLModel

class User(SQLModel, table=True):
    ...
    products: Mapped[list[Product]] = Relationship(
        sa_relationship=relationship(back_populates="user")
    )

class Product(SQLModel, table=True):
    ...
    user_id: int = Field(foreign_key="user.id")
    user: Mapped[User] = Relationship(
        sa_relationship=relationship(back_populates="products")
    )
```

**Reply by Arman-Hadi** [reply]
Really helped me. Thanks.
